File: api/src/test/java/kimosabe/api/service/UserServiceTest.java
package kimosabe.api.service;

import kimosabe.api.exceptions.IncorrectPasswordException;
import kimosabe.api.exceptions.MissingDatabaseEntryException;
import kimosabe.api.exceptions.UsernameTakenException;
import kimosabe.api.model.Role;
import kimosabe.api.model.RoleName;
import kimosabe.api.model.User;
import kimosabe.api.repository.RoleRepository;
import kimosabe.api.repository.UserRepository;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Assertions.*;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.Mockito;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.security.crypto.password.PasswordEncoder;
import static org.assertj.core.api.Assertions.assertThat;

import java.util.Optional;

import static org.mockito.Mockito.*;

@ExtendWith(MockitoExtension.class)
public class UserServiceTest {
    @Mock
    UserRepository userRepository;
    @Mock
    PasswordEncoder passwordEncoder;
    @Mock
    RoleRepository roleRepository;
    @InjectMocks
    UserService userService;

    private User user;

    @BeforeEach
    public void setUpHappyPath() {
        // Arrange
        user = new User("username", "password");
        when(passwordEncoder.encode("password")).thenReturn("encodedPassword");
        when(roleRepository.findByName(RoleName.ROLE_USER)).thenReturn(Optional.of(new Role()));
        when(userRepository.save(user)).thenReturn(user);
        when(userRepository.findByUsername("username")).thenReturn(Optional.empty());
    }

    @AfterEach
    public void cleanUp() {
        Mockito.reset(userRepository, passwordEncoder, roleRepository);
    }

    @Test
    public void whenCreateNewUserCalledValid_ThenNewUserShouldBeSaved() throws Exception {
        // Act
        User u = userService.createNewUser(user);

        // Assert
        verify(userRepository, times(1)).save(user);
        assertThat(u).isEqualTo(user);
    }

    @Test
    public void whenCreateNewUserCalledValid_ThenPasswordShouldBeEncrypted() throws Exception {
        // Act
        User u = userService.createNewUser(user);

        // Assert
        verify(passwordEncoder, times(1)).encode("password");
    }

    @Test
    public void whenCreateNewUserCalledUsernameTaken_ThenExceptionThrown() {
        // Arrange
        when(userRepository.findByUsername("username")).thenReturn(Optional.of(new User("username", "password")));

        // Act
        // Assert
        Assertions.assertThrows(UsernameTakenException.class, () -> userService.createNewUser(user));
    }

    @Test
    public void whenCreateNewUserCalledRoleMissing_ThenExceptionThrown() {
        // Arrange
        when(roleRepository.findByName(RoleName.ROLE_USER)).thenReturn(Optional.empty());

        // Act
        // Assert
        Assertions.assertThrows(MissingDatabaseEntryException.class, () -> userService.createNewUser(user));
    }

    @Test
    public void whenLoadUserByUsernameCalledInvalid_ThenThrowException() {
        // Act
        // Assert
        Assertions.assertThrows(UsernameNotFoundException.class, () -> userService.loadUserByUsername("username"));
    }

    @Test
    public void whenCheckPasswordCalledValid_ThenReturnTrue(){
        // Arrange
        when(passwordEncoder.matches("password", "password")).thenReturn(true);

        // Act
        // Assert
        Assertions.assertEquals(true, userService.checkPassword(user, "password"));
    }

    @Test
    public void whenCheckPasswordCalledIncorrectOldPassword_ThenReturnFalse(){
        // Assert
        Assertions.assertEquals(false, userService.checkPassword(user, "wrongPassword"));
    }

    @Test
    public void whenChangePasswordCalled_ThenPasswordChanged(){
        // Arrange
        when(passwordEncoder.encode("newPassword")).thenReturn("encodedNewPassword");

        // Act
        userService.changePassword(user, "newPassword");

        // Assert
        assertThat(user.getPassword()).isEqualTo("encodedNewPassword");
    }
}

File: api/src/test/kotlin/com/kimosabe/api/service/UserServiceTest.kt
package com.kimosabe.api.service

import com.kimosabe.api.model.Role
import com.kimosabe.api.model.RoleName
import com.kimosabe.api.model.User
import com.kimosabe.api.repository.RoleRepository
import com.kimosabe.api.repository.UserRepository
import io.mockk.clearAllMocks
import io.mockk.every
import io.mockk.impl.annotations.InjectMockKs
import io.mockk.impl.annotations.MockK
import io.mockk.junit5.MockKExtension
import io.mockk.verify
import io.mockk.verifyAll
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import org.junit.jupiter.api.extension.ExtendWith
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.test.context.ActiveProfiles
import java.util.*

@ExtendWith(MockKExtension::class)
class UserServiceTest {
  @MockK
  private lateinit var userRepository: UserRepository
  @MockK
  private lateinit var roleRepository: RoleRepository
  @MockK
  private lateinit var passwordEncoder: PasswordEncoder

  @InjectMockKs
  private lateinit var userService: UserService

  @AfterEach
  fun cleanup() {
    clearAllMocks()
  }

  @Test
  fun whenCreateNewUserCalled_ThenNewUserShouldBeAdded() {
    // Arrange
    val user : User = User(UUID.randomUUID(), "username", "password")
    every { passwordEncoder.encode("password") } returns "encodedpassword"
    every { roleRepository.findByName(RoleName.ROLE_USER) } returns Role()
    every { userRepository.save(user) } returns user

    // Act
    userService.createNewUser(user)

    // Assert
    verifyAll {
      passwordEncoder.encode("password")
      roleRepository.findByName(RoleName.ROLE_USER)
    }
    verify(exactly=1) { userRepository.save(user) }
  }

  @Test
  fun whenCreateNewUserCalled_ThenPasswordShouldBeEncrypted() {
    // Arrange
    val user : User = User(UUID.randomUUID(), "username", "password")
    every { passwordEncoder.encode("password") } returns "encodedpassword"
    every { roleRepository.findByName(RoleName.ROLE_USER) } returns Role()
    every { userRepository.save(user) } returns user

    // Act
    val response = userService.createNewUser(user)

    // Assert
    assertThat(response.password).isEqualTo("encodedpassword")
  }

  @Test
  fun whenLoadUserByUsernameCalled_AndUsernameIsValid_ThenReturnUser() {
    // Arrange
    val user : User = User(UUID.randomUUID(), "username", "password")
    every { userRepository.findByUsername("username") } returns user

    // Act
    val response = userService.loadUserByUsername("username")

    // Assert
    verifyAll {
      userRepository.findByUsername("username")
    }
    assertThat(response.password).isEqualTo("password")
  }

  @Test
  fun whenLoadUserByUsernameCalled_AndUsernameIsInalid_ThenErrorShouldBeThrown() {
    // Arrange
    every { userRepository.findByUsername("username") } returns null

    // Assert
    assertThrows<Error> { userService.loadUserByUsername("username") }
    verifyAll {
      userRepository.findByUsername("username")
    }
  }
}